pkg/protocol/sybilprotection/sybilprotectionv1/performance/performance.go

```diff
@@ -144,46 +188,90 @@ func (t *Tracker) ApplyEpoch(epoch iotago.EpochIndex, committee *account.Account
 	t.latestAppliedEpoch = epoch
 }
 
-func (t *Tracker) EligibleValidatorCandidates(_ iotago.EpochIndex) ds.Set[iotago.AccountID] {
-	// TODO: we should choose candidates we tracked performance for, only active
+// aggregatePerformanceFactors calculates epoch performance factor of a validator based on its performance in each slot by summing up all active subslots.
+func (t *Tracker) aggregatePerformanceFactors(slotActivityVector []*model.ValidatorPerformance, epoch iotago.EpochIndex) uint64 {
+	if len(slotActivityVector) == 0 {
+		return 0
+	}
 
-	return ds.NewSet[iotago.AccountID]()
+	var epochPerformanceFactor uint64
+	for _, pf := range slotActivityVector {
+		// no activity in a slot
+		if pf == nil {
+			continue
+		}
+		// each one bit represents at least one block issued in that subslot,
+		// we reward not only total number of blocks issued, but also regularity based on block timestamp
+		slotPerformanceFactor := bits.OnesCount32(pf.SlotActivityVector)
+
+		if pf.BlockIssuedCount > t.apiProvider.APIForEpoch(epoch).ProtocolParameters().RewardsParameters().ValidatorBlocksPerSlot {
+			// we harshly punish validators that issue any blocks more than allowed
+
+			return 0
+		}
+		epochPerformanceFactor += uint64(slotPerformanceFactor)
+	}
+
+	return epochPerformanceFactor
```

Comment:
```suggestion
	return epochPerformanceFactor >> SlotsPerEpochExponent
```
no?

Reply:
actually, maybe that's not the best place to add this `SlotsPerEpochExponent`; maybe it's best to do this shift when we use the performance factor and divide it by `validatorBlocksPerSlot`. Anyway, I don't see the `SlotsPerEpochExponent` being accounted for anywhere in the code, am I missing something?

Reply:
I've applied your suggestion; indeed `SlotsPerEpochExponent` wasn't used anywhere except when subtracted to the generationRateExponent to shift to the final result of the rewards https://github.com/iotaledger/iota.go/blob/81e2dc34386938e2fbd862ce3d21800470f38df5/mana.go#L69

Reply:
This test now fails https://github.com/iotaledger/iota-core/blob/178f30e00dff24b4581250e6471a781dfe714e00/pkg/protocol/sybilprotection/sybilprotectionv1/performance/tracker_test.go#L9

Reply:
I found the reason why the test was failing after this suggestion, we have to change something in testsuite_test.go